### ISSUE TITLE
Add support for here-doc

### ISSUE THREAD

**Issue by Tishj** [42-Ikole-Systems/TMK-SH#21]
https://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_07_04

It looks like this should be treated as a stack
```
cat <<eof1; cat <<eof2
Hi,
eof1
Helene.
eof2
```

Judging from the Shell Grammar rules, it seems this can not be determined by the Lexer, and only the Parser can determine whether a here-doc start (`io_here`) has been located.
```
io_here          : DLESS     here_end
                 | DLESSDASH here_end
                 ;
here_end         : WORD                      /* Apply rule 3 */
```

**Comment by Tishj** [comment]
When processing a given line, we need to likely keep a queue of here-doc tokens.
Once the next newline token has been processed, we need to start processing the next line(s) as the here-doc.

I don't think it makes sense to inform the Lexer of this, instead we want to directly consume lines from the same provider as the Lexer.
We can keep the here-doc lexing logic out of the Lexer, and not even need to return to the Lexer to fill in the content of the `io_here` symbol.

**Comment by mraasvel** [comment]
I agree with leaving it out of the lexer. It looks like it's intended to be done by the parser.

I believe the behavior on what to do with multiple here-docs is not documented. ZSH does appending, bash just takes the last one, so we have some freedom here.

**Comment by Tishj** [comment]
It is documented, it should be treated as a queue, the first here-doc gets processed first.
> If more than one "<<" or "<<-" operator is specified on a line, the here-document associated with the first operator shall be supplied first by the application and shall be read first by the shell.

> The here-document shall be treated as a single word that begins after the next <newline> and continues until there is a line containing only the delimiter and a <newline>, with no <blank>s in between. Then the next here-document starts, if there is one. The format is as follows:

First process the first, then the second, then the third etc..

**Comment by mraasvel** [comment]
I was talking about what to do with the output, but that might be mentioned in redirection now that I think about it

**Comment by Tishj** [comment]
Not what you were asking, but important to note here:
> Quote removal shall be applied to the word to determine the delimiter that is used to find the end of the here-document that begins after the next <newline>.

**Comment by Tishj** [comment]
And where the output goes is described here:
> io_redirect      :           io_file
                 | IO_NUMBER io_file
                 |           io_here
                 | IO_NUMBER io_here

It's just a variation of the `io_redirect` node
